### ISSUE TITLE
[question] Revisions and requirements overriding

### ISSUE THREAD

**Issue by roszakg** [conan-io/conan#13938]
### What is your question?

Hello everyone,

I have a question regarding the recipe revisions and support for requirement overriding.

I have two different recipe revisions hosted in local Artifactory and the only difference between the recipes is the requirements specification (they build the same version of the library - only the dependency is different):

- Recipe one:
```
def requirements(self):
       ...
       if self.options.with_openssl:
           self.requires("openssl/1.1.1p")
       ...
```
  

- Recipe two:
```
def requirements(self):
       ...
       if self.options.with_openssl:
           self.requires("openssl/1.1.1t")
       ...
```

I was expecting that depending on the explicit requirement override the proper revision of the binary package will be pulled into cache. Unfortunately this is not the case and even if openssl/1.1.1p is explicitly requested the binary created from the recipe two (the newer revision) is being used with the following warning:

```
cassandra-cpp-driver/2.16.0: Downloaded recipe revision 7442e6ce8e6a7431a8d2191fa83fc71d
WARN: cassandra-cpp-driver/2.16.0: requirement openssl/1.1.1t overridden by your conanfile to openssl/1.1.1p 
```

This creates a huge problem where older branches of the product I'm working on are actually using (or trying to use) the never version of the recipe which was build with the different binary dependencies. 

What is the proper way of handling cases like that? 
I wanted to avoid adding explicit suffix versioning to the 3rd party library.

To summarize - when I'm updating some dependency (like OpenSSL) I want to update all the packages that were using this dependency, but without breaking the old branches that were using those same packages just built with different versions of the package they depend on.

Also is there any way aside from rewriting Git history and adding explicit revision specification to the older branches to get out of the mess I have put myself into?

Thank you and kind regards,
Grzegorz 
 

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @roszakg 

Thanks for your question.
I am afraid that I don't fully understand the issue.

On one hand, there is an override somewhere, and even if ``cassandra-cpp-driver/2.16.0`` depends on ``openssl/1.1.1t``, that is being overriden by the downstream. This is related to the "graph computation" stage, in which the full graph of versions and recipe-revisions is computed. But not sure about this override, I guess some consuming package is defining such override. 

On the other hand, it seems there are also some questions related to the binary packages, the computed ``package_id`` for consumers, that happens later in the "binary computation" stage, that happens only after the full graph has been computed.

You mention ``but without breaking the old branches``, what that does mean? You mean previous packages? It is not very clear what "breaking the old branches" mean, what is the breakage?

Maybe also openssl/1 is a bad case, because it is not a good candidate for version ranges, at least with Conan 1 (Conan 2.0 allows version ranges better over openssl/1).

By the way, you don't mention the Conan version. It is very important to clarify which one, now that Conan 2.0 is out and that is the latest one.

If you could please try to clarify a bit better the above points, that would help. Thanks!

**Comment by roszakg** [comment]
Hi @memsharded,

Let's start with Conan version - I'm still using Conan 1.x (mainly due to the fact that we are using SCons generator).

Please let me give you a specific example.

When I started I created the recipe for cassandra-cpp-driver/2.16.0 that was referencing openssl/1.1.1p and from that recipe I created the first revision (e5f9ade330133d0d94236d1254d68721) and two binary packages (with different ABI settings). Those were uploaded to Artifactory:

![image](https://github.com/conan-io/conan/assets/2522702/f6b4b6ee-01f3-4bf1-88e7-1e1e40de1b66)

At that point my conanfile.txt looked like that:
```
...
cassandra-cpp-driver/2.16.0
openssl/1.1.1p
...
```

After releasing few versions of our software (and creating branches for those releases) we decided that it is time to upgrade the openssl to the newer version (openssl/1.1.1t), but instead of only updating openssl requirement in conanfile.txt we also created the new package revision for the cassandra-cpp-driver/2.16.0 with modified "def requirements() method". 
We uploaded this new revision (7442e6ce8e6a7431a8d2191fa83fc71d) and new binary to the Artifactory:
![image](https://github.com/conan-io/conan/assets/2522702/2d096e89-089f-4a06-951b-35e67d6b5444)

After that we modified our conanfile.txt on the master branch to:
```
...
cassandra-cpp-driver/2.16.0
openssl/1.1.1t
...
```

Everything worked fine - the new revision of the package was pulled in and the code on the master branch worked just fine.

However when we checkout one of the older branches (that one that was using the openssl/1.1.1p) then instead of the older revision of cassandra-cpp-driver (e5f9ade330133d0d94236d1254d68721) the newer one (7442e6ce8e6a7431a8d2191fa83fc71d) is being used and this is causing problems.

What I'm looking for is a way to force coupling between the requirements and the specific revision. 
Since the explicit requirements from the conanfile.txt are in conflict with specific package revision I guess I was expecting Conan to find the better matching revision of that package automatically.

I hope that clarifies a little bit the problem we are facing.

Thanks,
Grzegorz

**Comment by memsharded** [comment]
Thanks for the clarifications @roszakg 

I'd say that you are facing a common situation induced by not locked dependencies, or achieving deterministic dependency graphs over time when such dependencies keep evolving.

In essence, revisions are an automatic extra version number (like ``cassandra-cpp-driver/2.16.0.revision``) that automatically resolves by default to the latest existing revision.

There are different approaches to this situation (those are common in package management, not only in Conan):

1. When you do a modification to a package like ``cassandra``, even if it is just bumping a dependency version, you still change the ``cassandra`` version let's say to ``cassandra/2.16.0.1``. That way the old branches will still depend on the previous ``cassandra/2.16.0`` and nothing will change.
2. If you rely on revisions to avoid explicit bumping in conanfiles, you can still reference exact revisions in your consumer ``conanfile.txt`` something like:
    ```
    [requires]
    cassandra-cpp-driver/2.16.0#e5f9ade330133d0d94236d1254d68721
    openssl/1.1.1t#revision
    ```
    This approach might partially work, but if there are intermediate transitive dependencies, like ``cassandra`` depends also on ``boost/1.80`` and ``zlib/1.2.13``, that are not directly defined in ``conanfile.txt``, those dependencies might still change over time. Note that it is also possible to specify ``conanfile.py`` dependencies down to the recipe revision with ``requires = "openssl/1.1.1t#reciperevision"``.

3. Use lockfiles: https://docs.conan.io/2/tutorial/consuming_packages/intro_to_versioning.html#lockfiles and https://docs.conan.io/2/tutorial/versioning.html
  Lockfiles are a common mechanism in package managers. It is an auxiliary file that captures a snapshot of the dependencies versions and revisions, to be able to resolve to that same set of versions and revisions later in time, and achieve reproducibility of the dependency graph, when there are moving pieces (if specifying revisions explicitly in conanfiles, there will be no moving pieces). A note about lockfiles: they can be challenging to master and use in Conan 1.X, but they have been greatly simplified in 2.0. If you are not using lockfiles yet, I'd recommended migrating first to Conan 2.0, then learn the lockfiles in 2.0.

Which approach to use depends on the specific needs. All are trade-offs of control, flexibility and speed of evolution. If all conanfiles hardcode the revisions in the ``requires = "dep/version#revision"``, the control is total and the reproducibility is automatic, but it can be very cumbersome and tedious to evolve it when new revisions are created. On the other hand, using lockfiles typically allows to iterate faster, to test new revisions faster without modifications to the conanfiles, but in order to control the reproducibility of previous branches, it is necessary to store the lockfiles associated with those branches.

**Comment by memsharded** [comment]
It seems that this question didn't have any follow ups or comments.
I am closing it as answered back then, but if there are any further question or issues, please comment or create new tickets as necessary. Thanks for the feedback!
